src/python/pants/backend/codegen/tasks/simple_codegen_task.py

```diff
@@ -0,0 +1,155 @@
+# coding=utf-8
+# Copyright 2015 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+
+from pants.backend.core.tasks.task import Task
+from pants.base.address import SyntheticAddress
+from pants.base.build_environment import get_buildroot
+
+
+class SimpleCodegenTask(Task):
+  """Simpler base-class for single-language code-gen.
+
+  Subclasses should implement at minimum: synthetic_target_type, is_gentarget, execute_codegen, and
+  sources_generated_by_target.
+  """
+
+  @classmethod
+  def get_fingerprint_strategy(cls):
+    return None
+
+  def synthetic_target_extra_dependencies(self, target):
+    """Gets any extra dependencies generated synthetic targets should have. This method is optional
+    for subclasses to implement, because some code generators may have no extra dependencies.
+
+    :param target: the Target from which we are generating a synthetic Target. E.g., 'target' might
+    be a JavaProtobufLibrary, whose corresponding synthetic Target would be a JavaLibrary. It is not
+    necessary to use this parameter; it may be unnecessary depending on the details of the subclass.
+    :return: a list of dependencies.
+    """
+    return []
+
+  @property
+  def synthetic_target_type(self):
+    """The type of target this codegen task generates. For example, the target type for JaxbGen
+    would simply be JavaLibrary.
+
+    :return: a type (class) that inherits from Target.
+    """
+    raise NotImplementedError
+
+  def is_gentarget(self, target):
+    """Predicate which determines whether the target in question is relevant to this codegen task.
+    E.g., the JaxbGen task considers JaxbLibrary targets to be relevant, and nothing else.
+
+    :param target: The target to check.
+    :return: True if this class can generate code for the given target, False otherwise.
+    """
+    raise NotImplementedError
+
+  def execute_codegen(self, invalid_targets):
+    """Generated code for the given list of targets.
+
+    :param invalid_targets: an iterable of targets (a subset of codegen_targets()).
+    """
+    raise NotImplementedError
+
+  def sources_generated_by_target(self, target):
```

Comment:
Yeah; it would be easy with strategy=isolated. 

In the global case, I've also thought about just checking the diff of the ./pants.d/gen/[wire/jaxb/etc] folder, but that might slow things down a bit. I could try it. I'd worry though for the global case, that two different targets might generate the exact same files, and the diff wouldn't pick up on that. This would result in the second target's dependees not picking those generated sources up as dependencies, even when they should.